src/network/mod.rs

```diff
@@ -59,45 +102,89 @@ impl<C: Codec, M: MultihashDigest> Network<C, M> {
-                Poll::Ready(Some(event)) => event,
+            match Pin::new(&mut self.rx).poll_next(ctx) {
+                Poll::Ready(Some((cid, tx))) => {
+                    let entry = self.wanted.entry(cid.clone()).or_default();
+                    entry.add_receiver(tx);
+                    self.swarm.want_block(cid, 1000);
+                }
                 Poll::Ready(None) => return Poll::Ready(()),
                 Poll::Pending => break,
-            };
-            log::trace!("{:?}", event);
-            match event {
-                StorageEvent::Want(cid) => self.swarm.want_block(cid, 1000),
-                StorageEvent::Cancel(cid) => self.swarm.cancel_block(&cid),
-                StorageEvent::Provide(cid) => {
-                    if let Err(err) = match self.storage.get_local(&cid) {
-                        Ok(Some(block)) => self.swarm.provide_and_send_block(&cid, &block),
-                        _ => self.swarm.provide_block(&cid),
-                    } {
-                        log::error!("error providing block {:?}", err);
+            }
+        }
+
+        if self.bootstrap_complete {
+            loop {
+                let event = match Pin::new(&mut self.subscription).poll_next(ctx) {
+                    Poll::Ready(Some(event)) => event,
+                    Poll::Ready(None) => return Poll::Ready(()),
+                    Poll::Pending => break,
+                };
+                log::trace!("{:?}", event);
+                match event {
+                    Event::Insert(cid) => {
+                        if let Err(err) = match self.blocks.get(&cid) {
+                            Ok(Some(data)) => self.swarm.provide_and_send_block(&cid, &data),
+                            _ => self.swarm.provide_block(&cid),
+                        } {
+                            log::error!(
+                                "{}: error providing block {:?}",
+                                self.swarm.node_name(),
+                                err
+                            );
+                        }
                     }
+                    Event::Remove(cid) => self.swarm.unprovide_block(&cid),
                 }
-                StorageEvent::Unprovide(cid) => self.swarm.unprovide_block(&cid),
             }
         }
+
+        loop {
+            match Pin::new(&mut self.interval).poll_next(ctx) {
+                Poll::Ready(Some(())) => {}
+                Poll::Ready(None) => return Poll::Ready(()),
+                Poll::Pending => break,
+            }
+            let timedout = Instant::now() - self.timeout;
+            let mut wanted = std::mem::replace(&mut self.wanted, HashMap::with_capacity(0));
+            wanted.retain(|cid, wanted| {
+                if wanted.timestamp > timedout {
+                    true
+                } else {
+                    self.swarm.cancel_block(cid);
+                    false
+                }
```

Comment:
Something about this feels off to me. To check if the request has timed out, it checks if the timestamp on the request is less than or equal to the difference between `Instant::now()` and `self.timeout`. If it is less than the difference or equal to the difference, it cancels the request. If it is greater than the difference, it retains the request.

Reply:
It's legit, it's tested in the exchange dht test when the waiting period for the bootstrap node is too small